### PR TITLE
feat: Added SMTP extension

### LINKED ISSUE #158: SMTP extension
Add a new core extension to allow Chainloop users to send email notifications for every workflow run. 
This will also be an excellent opportunity to validate a new experimental extensions API.

### PR DESCRIPTION
**Pull request by danlishka**
This is a new extension for sending emails via SMTP. This is still a very early experimental version focusing on validating the latest cutting-edge integrations framework.

The README file provides more information on how to use this extension.

 closes #158

### CODE REVIEW

File: app/controlplane/extensions/core/smtp/v1/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+## How to use it
+
+In the following example, we will use the AWS SES service.
+
+1. To get started, you need to register the extension in your Chainloop organization.
+```
+chainloop integration add smtp --options user=AHDHSYEE7e73,password=kjsdfda8asd****,host=email-smtp.us-east-1.amazonaws.com,port=587,to=platform-team@example.com,from=notifier@example.com
+```
+
+2. When attaching the integration to your workflow, you have the option to specify CC:
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/README.md:14]: nitpick. I'd put two commands here, one without the option and one with the option to make clear `cc` is optional.

File: app/controlplane/extensions/core/smtp/v1/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# SMTP Fan-out Extension
+
+With this extension, you can send an email for every workflow run and attestation, ensuring better communication.
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/README.md:3]: > ensuring better communication.

not sure what you mean by this.

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+type Integration struct {
+	*sdk.FanOutIntegration
+}
+
+type registrationRequest struct {
+	To       string `json:"to" jsonschema:"format=email,description=The email address to send the email to."`
+	From     string `json:"from" jsonschema:"format=email,description=The email address of the sender."`
+	User     string `json:"user" jsonschema:"minLength=1,description=The username to use for the SMTP authentication."`
+	Password string `json:"password" jsonschema:"description=The password to use for the SMTP authentication."`
+	Host     string `json:"host" jsonschema:"description=The host to use for the SMTP authentication."`
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:39]: Would any of these format help? https://json-schema.org/understanding-json-schema/reference/string.html#hostnames

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+	Port     string `json:"port" jsonschema:"description=The port to use for the SMTP authentication"`
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:40]: I am assuming you are using a string here because of this bug https://github.com/chainloop-dev/chainloop/issues/157

Maybe it would make sense to add a comment here so we do not forget to change it?

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+type registrationState struct {
+	To   string `json:"to"`
+	From string `json:"from"`
+	User string `json:"user"`
+	Host string `json:"host"`
+	Port string `json:"port"`
+}
+
+type attachmentRequest struct {
+	Cc string `json:"cc,omitempty" jsonschema:"format=email,description=The email address of the carbon copy recipient."`
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:52]: nitpick, acronyms should be all in capital letters. https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+type attachmentState struct {
+	Cc string `json:"cc"`
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:56]: same uppercase comment here.

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+func New(l log.Logger) (sdk.FanOut, error) {
+	base, err := sdk.NewFanOut(
+		&sdk.NewParams{
+			ID:      "smtp",
+			Version: "0.1",
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:63]: good, we'll bump down the other integrations too.

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+	// Unmarshal the request
+	var request *registrationRequest
+	if err := sdk.FromConfig(req.Payload, &request); err != nil {
+		return nil, fmt.Errorf("invalid registration request: %w", err)
+	}
+
+	response := &sdk.RegistrationResponse{}
+
+	to, from, user, password, host, port := request.To, request.From, request.User, request.Password, request.Host, request.Port
+	rawConfig, err := sdk.ToConfig(&registrationState{To: to, From: from, User: user, Host: host, Port: port})
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:93]: I'd move the code about generating the response (and state) after the main registration logic

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("marshalling configuration: %w", err)
+	}
+	response.Configuration = rawConfig
+	response.Credentials = &sdk.Credentials{Password: password}
+
+	// validate and notify
+	subject := "[chainloop] New SMTP integration added!"
+	tpl := `
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:102]: It would be ideal to use go-templates instead https://pkg.go.dev/text/template, it can be in another patch.

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+	to, from, user, password, host, port := request.To, request.From, request.User, request.Password, request.Host, request.Port
+	rawConfig, err := sdk.ToConfig(&registrationState{To: to, From: from, User: user, Host: host, Port: port})
+	if err != nil {
+		return nil, fmt.Errorf("marshalling configuration: %w", err)
+	}
+	response.Configuration = rawConfig
+	response.Credentials = &sdk.Credentials{Password: password}
+
+	// validate and notify
+	subject := "[chainloop] New SMTP integration added!"
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:101]: ok, so this is in fact sending an email during registration. That's an interesting approach. If we just wanted to check that the credentials are correct, is there any other way to check it?

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+	body := fmt.Sprintf(tpl, req.WorkflowID, jsonBytes, i.Describe().ID, i.Describe().Version)
+	err = sendEmail(host, port, user, password, from, to, ac.Cc, subject, body)
+	if err != nil {
+		return fmt.Errorf("sending an email: %w", err)
+	}
+
+	return nil
+}
+
+func validateExecuteRequest(req *sdk.ExecutionRequest) error {
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:201]: mental note for myself: this is starting to look more and more like smth that could get automatically validated.

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+	message := "From: " + from + "\n" +
+		"To: " + to + "\n" +
+		"Cc: " + cc + "\n" +
+		"Subject: " + subject + "\n\n" +
+		body
+
+	auth := nsmtp.PlainAuth("", user, password, host)
+	err := nsmtp.SendMail(host+":"+port, auth, from, []string{to}, []byte(message))
+	if err != nil {
+		fmt.Println("Error sending email1:", err)
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:231]: there is a namespaced logger instance injected (see dependency track example), use it instead.

File: app/controlplane/extensions/core/smtp/v1/extension_test.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+		errMsg string
+	}{
+		{
+			name:   "missing properties",
+			input:  map[string]interface{}{},
+			errMsg: "missing properties: 'to', 'from', 'user'",
+		},
+		{
+			name:  "valid request",
+			input: map[string]interface{}{"from": "test@example.com", "to": "test@example.com", "host": "localhost", "port": "25", "user": "test", "password": "test"},
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension_test.go:39]: probably test email formats and potential hostnames.

File: app/controlplane/extensions/core/smtp/v1/extension.go [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+	}
+
+	if req.AttachmentInfo == nil || req.AttachmentInfo.Configuration == nil {
+		return errors.New("missing attachment configuration")
+	}
+
+	return nil
+}
+
+func sendEmail(host string, port string, user, password, from, to, cc, subject, body string) error {
```
Comment (migmartri) [inline app/controlplane/extensions/core/smtp/v1/extension.go:221]: when you start having more than 2 arguments, specially if they are from the same type, it's common to change the function signature to accept a struct instead. That makes it more readable and less error prone. I'd suggest doing that here

**Review by migmartri — COMMENTED** [review]
Awesome, good start! 

I left some comments, most of them suggestions that can be changed in another patch.

For example, by encapsulating the render code in a function + go templates you'll be able to test them.

I'd suggest to also add comments in the code explaining/separating the different blocks. And as a general rule, make sure that initialization of variables/code happens the closest to the code that'll use it as possible (i.e preparing the response in `register`.

Looking good!

File: app/controlplane/extensions/core/smtp/v1/extension.go (hunk 3 above)
Comment (danlishka) [reply app/controlplane/extensions/core/smtp/v1/extension.go:39]: This is not supported by our jsonschema library - I can work on the PR though. We would have to use something like: `anyof_format=hostname,ipv4`

to get

```
"anyOf": [
{ "format": "hostname" },
{ "format": "ipv4" },
{ "format": "ipv6" }
]
```

File: app/controlplane/extensions/core/smtp/v1/extension.go (hunk 10 above)
Comment (danlishka) [reply app/controlplane/extensions/core/smtp/v1/extension.go:101]: Yes, this way, you really make sure you receive an email. It might still go to spam, or the `to` address might be incorrect, or the `from` address could be rejected by the SMTP service, etc.

File: app/controlplane/extensions/core/smtp/v1/extension.go (hunk 9 above)
Comment (danlishka) [reply app/controlplane/extensions/core/smtp/v1/extension.go:102]: Yes, the SMTP logic will also be moved to a separate package.

File: app/controlplane/extensions/core/smtp/v1/extension.go (hunk 14 above)
Comment (danlishka) [reply app/controlplane/extensions/core/smtp/v1/extension.go:221]: yes, this will change once we move it out to a separate package

**Comment by migmartri** [comment]
Please link the issue in the description not in the title of the PR.

**Review by migmartri — APPROVED** [review]
LGTM!
